src/main/java/com/mraof/minestuck/tileentity/redstone/StructureCoreTileEntity.java

```diff
@@ -0,0 +1,256 @@
+					return type;
+			}
+			throw new IllegalArgumentException("Invalid ordinal of " + ordinal + " for structure core action type!");
+		}
+		
+		public String getNameNoSpaces()
+		{
+			return name().replace('_', ' ');
+		}
+	}
+	
+	public StructureCoreTileEntity()
+	{
+		super(MSTileEntityTypes.STRUCTURE_CORE.get());
+		actionType = ActionType.READ_AND_REDSTONE;
+	}
+	
+	@Override
+	public void tick()
+	{
+		if(level == null || level.isClientSide)
+			return;
+		
+		int cycleRate = (hasWiped && actionType == ActionType.READ_AND_WIPE) ? 600 : 100;
+		if(tickCycle >= cycleRate)
+		{
+			sendUpdate();
+			tickCycle = 0;
+		}
+		tickCycle++;
+	}
+	
+	private void sendUpdate()
+	{
+		if(level != null && level.isAreaLoaded(getBlockPos(), 1) && getBlockState().getValue(StructureCoreBlock.ACTIVE))
+		{
+			StructureManager structureManager = ((ServerWorld) level).structureFeatureManager();
+			StructureStart<?> structureStart = getStructureStart(structureManager);
+			if(structureStart != null && structureStart.isValid())
+			{
+				Debug.debugf("meets all conditions, now doing action type function");
+				if(actionType == ActionType.WRITE)
+				{
+					writeToStructure(structureStart);
+				} else if(actionType == ActionType.READ_AND_WIPE)
+				{
+					wipeSlate();
+				} else if(actionType == ActionType.READ_AND_REDSTONE)
+				{
+					CoreCompatibleScatteredStructurePiece piece = getStructurePiece(structureStart);
+					
+					if(piece != null && piece.hasBeenCompleted())
+					{
+						if(!getBlockState().getValue(StructureCoreBlock.POWERED))
+							level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, true));
+					} else
+					{
+						level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, false));
+					}
+				}
+			}
+		}
+	}
+	
+	public StructureStart<?> getStructureStart(StructureManager structureManager)
+	{
+		for(Structure<?> structureListIterate : SOLVABLE_STRUCTURES)
+		{
+			return structureManager.getStructureAt(getBlockPos(), true, structureListIterate);
+		}
+		
+		return null;
+	}
+	
+	private void writeToStructure(StructureStart<?> structureStart)
+	{
+		//compoundnbt.putString("id", Registry.STRUCTURE_FEATURE.getKey(this.getFeature()).toString());
+		
+		CoreCompatibleScatteredStructurePiece piece = getStructurePiece(structureStart);
+		if(piece != null)
+		{
+			if(!piece.hasBeenCompleted())
+				piece.nowCompleted();
+			else
+				Debug.debugf("has already been completed");
+		}
+	}
+	
+	private CoreCompatibleScatteredStructurePiece getStructurePiece(StructureStart<?> structureStart)
+	{
+		List<StructurePiece> structurePieceList = structureStart.getPieces();
+		for(StructurePiece pieceIterate : structurePieceList)
+		{
+			if(pieceIterate instanceof CoreCompatibleScatteredStructurePiece)
+			{
+				return (CoreCompatibleScatteredStructurePiece) pieceIterate;
+			}
+		}
+		
+		return null;
+	}
+	
+	private void wipeSlate()
```

Comment:
The behavior of this function doesn't seem to match the javadoc for `READ_AND_WIPE`, as it doesn't check if the structure has been completed.

Reply:
Should be improved now, all functions related to a specific enum type have been reorganized and now include some of the same checks